### ISSUE TITLE
Harden train-step numerics: FP32 gradient accumulation and atomic nonfinite-step skipping

### ISSUE THREAD

**Issue by mlnomadpy** [mlnomadpy/flaxchat#6]
Parent: #1

Observed on `master` at commit `9653ab4`.

## Problem

The accumulated train step initializes gradient buffers with:

```python
jax.tree.map(jnp.zeros_like, nnx.state(model, nnx.Param))
```

so the accumulation dtype follows the parameter dtype. With BF16 parameters, small gradient contributions can be lost across microbatches.

The current nonfinite handling replaces bad gradients with zeros and still calls `optimizer.update`. Stateful optimizers can therefore advance their count, moments, schedules, or decoupled weight decay even though the gradient step was invalid.

## Proposed change

Use a configurable accumulation dtype, defaulting to FP32 for BF16 training. Gate the complete optimizer transition with `lax.cond`: when loss or gradients are nonfinite, preserve model parameters and optimizer state exactly and record a skipped update.

## Acceptance criteria

- BF16 training accumulates gradients in FP32 by default, with an explicit configuration override.
- Accumulated gradients are cast only at the optimizer boundary required by the selected optimizer policy.
- An injected NaN or Inf leaves parameters, optimizer count, moments, and weight-decay state unchanged.
- Metrics distinguish microbatches processed, successful optimizer updates, and skipped updates.
- A deterministic test compares microbatch accumulation against a matched large-batch reference.
- Tests cover both the single-microbatch and scanned accumulation paths.

**Comment by mlnomadpy** [comment]
Resolved on current master. Accumulation defaults to FP32 with an explicit override; both single and scanned paths are tested against a matched large batch. Atomic nonfinite gating preserves parameters and the complete optimizer state, while run/checkpoint metrics separately track microbatches, successful updates, and skipped updates.
